### ISSUE TITLE
Alternative to case class generation?

### ISSUE THREAD

**Issue by cornerman** [cornerman/sloth#10]
How can we support serialization libraries without automatic support for case classes to use this library?

* Shapeless records?

Related https://github.com/cornerman/covenant/issues/3

**Comment by Poorva17** [comment]
We are now trying upickle instead of protobuf, but we still hit this issue, because even upickle does not allow for automatic case class ser/de. We do not understand if chameleon.ext.upickle is of any use because of the above issue.

**Comment by cornerman** [comment]
Thanks for the feedback. When I added upickle, I did not have the requirement of automatic derivation for case class serializers. So I did not really think about it. Apart from this library with its specific requirement, upickle support in chameleon still has its use. 

For upickle, I have opened an issue: https://github.com/lihaoyi/upickle/issues/229

Alternatively for json, you could use circe. Does that make sense for you?

As a side note: The alternative of shapeless records would still mean, that we might need to add support to some serialization libraries, that do not support shapeless out of the box.

**Comment by deterdw** [comment]
Instead of generating intermediate case classes for serialization, isn't it possible to define a trait like
```scala
trait Pickler[PickledType] {
  type FieldT
  def apply(fields: Seq[String,FieldT]): PickledType
}
```
with the addtional constraint that the implementing object must have a method
```scala
  def apply[A](a: A): FieldT
```
for every type `A` that you want to serialise? E.g. an example for uPickle,
```scala
object uPicklePickler extends Pickler[Js.Value] {
  type FieldT=Js.Value
  def apply(fields: Seq[String,FieldT]): Js.Value = Js.Obj(fields)
  def apply[A: Writes](a: A): Js.Value = writeJs[A](a)
}
```
This way, your macro can, instead of generating a case class, just output the code
```scala
pickler(
  "arg1"->pickler(arg1),
  "arg2"->pickler(arg2),
  ...
)
```
without any dependency on Shapeless.

(Code written from memory, so might contain errors, but I hope the gist is clear.)

**Comment by cornerman** [comment]
@deterdw Good idea! This will definitely work for json, but I am not sure about binary serialization (like boopickle or protobuf). Does it still work there to serialize each argument and then serialize the `Seq[(String,PickleType)]`? Binary serialization would then typically drop the names and take the order of the arguments.

**Comment by deterdw** [comment]
Sorry, I didn't get a notification about a response... I am assuming that most serialization libraries will either use a field-based AST (like JSON) or be order-based. You could also add indices to the calls, like
```scala
pickler(
  ("arg1",0,pickler(arg1)),
  ("arg2",1,pickler(arg2)),
  ...
)
```
if need be. I don't see how any automatic serialization system based on non-annotated case classes can do better, because you don't have more information than the constructor parameter names, types and order. You could even drop the requirement that there must be a single internal `FieldT` in the serializer - all that it must do, is be able to execute a call of the above structure. (It seems rather like compile time structural subtyping to me.)

For deserializing on the server side, if you want to keep serialization errors apart from execution errors, you could do something like
```scala
val arg1=pickler.unpack[Type1]("arg1",0)
val arg2=pickler.unpack[Type2]("arg2",1)
...
``` 
followed by the invocation of `method(arg1,arg2,...)`. This supports named-based, position-based and even streaming codecs (which can ignore whatever of the three isn't relevant).

**Comment by nafg** [comment]
Why not just use tuples instead of case classes?

**Comment by cornerman** [comment]
A tuple does not carry any information about field names. The payload for the request and responses are the arguments and return types of scala methods. We want to be able to serialize both with json or binary serialization. And arguments have a name and a type that you want to capture. With case classes we have the names of the arguments and the types. So serialization methods like, e.g., json can have field names.

**Comment by nafg** [comment]
Why is it a problem if the fields are named _1, _2, _3 etc.? Are you
supporting overloaded methods?

**Comment by cornerman** [comment]
@nafg We are not supporting overloaded methods automatically, though you can make them work specifying a path-annotation to your method. `@Path("alternative-name")`. Because every method in an Api trait is identified by the name of the Api trait and the method name (or any path annotation). Then we need to have a specific method signature to know how to deserialize the payload.

I like to support proper names in order to make the API usable from other language (and for potentially facading exisiting json apis). As of now, I can work with my APIs from another language by sending an http request with json payload corresponding to the parameter list of the method. It is nice to use real names here.

If we drop this requirements of names, we could go with something different, we could even just serialize to a list instead of a tuple. Though, a Seq or Tuple would still not suffice, because it loses all type information of the single members. For every method, we want to be able to serialize all arguments together in one go, so the serializer can work efficiently. That means we need some composite type that includes all arguments with their values and correct types. You could use shapeless with their `HList` and records to represent what we have. But I do not really want to pull in shapeless as a dependency (it was this way some time ago).

All in all, I realize that generating a case class for each method is not so desriable because of compile-speed and generally because it introduces kind of unnecessary types. What we could do: invent our own type for storing an argumentlist (very similar to an HList in shapeless):

```scala
sealed trait ArgList extends Product with Serializable {
  def ::[H](h : (String, H)) = ArgCons(h._1, h._2, this)
}
final case class ArgCons[H, +T <: ArgList](name: String, head: H, tail: T) extends ArgList
case object ArgNil extends ArgList
``` 

In our macro, we would make every method construct a `ArgList` corresponding to its own method signature. Then this `ArgList` can be serialized by a serializer of your choice and send to the backend via your configured transport. For this to work, we would to make each serializer work with `ArgList`, but at least that would be a possibility. But I am not sure. I am totally open for new ideas :)

Does that make sense to you?

**Comment by cornerman** [comment]
For another possible way to deal with this: autowire does serialize each argument individually and then gives you a `Map[String, PickleType]`. That works, but has the drawback that you often need to serialize twice, once each value and then once for the whole map. Additionally, I do not see how to make this work efficiently with binary serialization.

We could also try out what @deterdw (btw. sorry for taking so long) is proposing, which supports name-based and position-based serialization. We would probably serialize twice for this approach as well (each argument and then typically the whole structure), but maybe that it not too problematic.

**Comment by Dennis4b** [comment]
Is there a way currently (with the latest scalajs 1 versions) to use sloth without requiring full automatic derivation?

If I just import `boopickle.Default._` and let it do its thing, my fullopt-js output grows from 4 megabytes (using `autowire+upickle`) to over 8 megabytes, with the optimization taking nearly 10 times as long. (This is due to picklers being generated over and over, sometimes thousands of times for a basic type that's used everywhere)

On the JVM I don't really care about this duplication but it's obviously bad for the JS side.

I can get this back down mostly by adding implicit picklers for the most commonly used types and hoping they are picked up before the auto-derivation macros (which isn't all that easy depending on how types are imported), but it's not optimal and it's easy to forget. It'd be much easier to use `boopickle.DefaultBasic._` (i.e. only semi-automatic derivation) and just fix the compilation errors - except that I then don't have the necessary picklers for the sloth generated case classes.

Does anyone have a suggestion? I like the performance/size of boopickle compared to upickle, and i like the ergonomics of sloth compared to autowire.

The PR in #19 doesn'y apply to the latest release.

**Comment by cornerman** [comment]
@Dennis4b Agree on the problem with case class generation. For now, I have rebased the PR on the current master - so if you want, you can use it via jitpack. And make it work with the tuple serialization. This should reduce your problems of generated serializers.

I am not ready to merge the PR yet, because I would like to iterate on it again - to find a maybe better solution. Not sure yet.

Does that work for you?

**Comment by Dennis4b** [comment]
Thank you for the quick reply and porting the PR! It works perfectly.

I've gone and annotated all types (about 700) with the boopickle semi-automatic derivation macros, and switched to using tuple serialization. My full-opt-js is now even a good bit smaller than it was using `upickle`. 

I am very happy with this solution, thank you!

**Comment by cornerman** [comment]
Fixed with  #100
